{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import ctypes\n",
    "from scipy.stats import chi2\n",
    "from statsmodels.sandbox.stats.multicomp import fdrcorrection_twostage as fdr\n",
    "import os\n",
    "import argparse\n",
    "import time\n",
    "from iotools.dosageparser import DosageParser\n",
    "from iotools.readVCF import ReadVCF \n",
    "from scipy.optimize import minimize\n",
    "from scipy.special import erfinv\n",
    "#from inference.regulizer_optimizer import OptimizeRegularizer\n",
    "from inference.per_snp_optimiser import OptimizeRegularizer\n",
    "#from inference.perm_regulizer_optimizer import OptimizeRegularizer\n",
    "#from select_genes_per_snp import linear_reg\n",
    "import regstat\n",
    "import regnull\n",
    "import hist_qq\n",
    "#from qqplot import plot_hist_qq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def logml(X, Y, sigmabeta, sigmax):\n",
    "    nsnps = X.shape[0]\n",
    "    nsamples = X.shape[1]\n",
    "    sigmabeta2 = sigmabeta * sigmabeta\n",
    "    sigmax2    = np.square(sigmax)\n",
    "    \n",
    "    Yt = Y.T # shape N x G\n",
    "    Abase = sigmabeta2 * np.dot(Yt, Yt.T)\n",
    "    \n",
    "    totres = 0\n",
    "    for i in range(nsnps):\n",
    "        A = Abase.copy()\n",
    "        A[np.diag_indices(nsamples)] += sigmax2[i]\n",
    "        logdetA = np.linalg.slogdet(A)\n",
    "        invA = np.linalg.inv(A)\n",
    "        res = - 0.5 * ( nsamples * np.log(2 * np.pi) + logdetA[1] + np.dot(X[i, :].T, np.dot(invA, X[i, :])) )\n",
    "        totres += res\n",
    "    return totres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def permuted_dosage(expr, nsnp = 5000, fmin = 0.1, fmax = 0.9, maketest = False):\n",
    "\n",
    "    f = np.random.uniform(fmin, fmax, nsnp)\n",
    "    if maketest:\n",
    "        f = np.repeat(0.1, nsnp)\n",
    "    nsample = expr.shape[1]\n",
    "\n",
    "    dosage = np.zeros((nsnp, nsample))\n",
    "    for i in range(nsnp):\n",
    "        if maketest:\n",
    "            nfreq = np.array([[279,  54,   5]])[0]\n",
    "        else:\n",
    "            mafratios = np.array([(1 - f[i])**2, 2 * f[i] * (1 - f[i]), f[i]**2])\n",
    "            nfreq  = np.random.multinomial(nsample, mafratios, size=1)[0]\n",
    "        f1 = np.repeat(0, nfreq[0])\n",
    "        f2 = np.repeat(1, nfreq[1])\n",
    "        f3 = np.repeat(2, nfreq[2])\n",
    "        x  = np.concatenate((f1,f2,f3))\n",
    "        dosage[i, :] = np.random.permutation(x)\n",
    "\n",
    "    maf2d = f.reshape(-1, 1)\n",
    "    gtnorm = (dosage - (2 * maf2d)) / np.sqrt(2 * maf2d * (1 - maf2d))\n",
    "    gtcent = dosage - np.mean(dosage, axis = 1).reshape(-1, 1)\n",
    "\n",
    "    return gtnorm, gtcent\n",
    "\n",
    "\n",
    "def Expression(gt, gx, beta, cfrac = 0.001):\n",
    "\n",
    "    ntrans  = gt.shape[0]\n",
    "    ngene   = gx.shape[0]\n",
    "    nsample = gx.shape[1]\n",
    "\n",
    "    liabilities = np.zeros((ngene, nsample))\n",
    "    cindex = np.zeros((ntrans, ngene))                                           # Index matrix of gene / trans-eQTLs pairs\n",
    "    nc  = np.random.gamma(ngene * cfrac, scale = 1.0, size = ntrans).astype(int) # number of target genes for each trans-eQTL\n",
    "    for i in range(ntrans):\n",
    "        ncausal = min(ngene, nc[i])                                              # do something better, trans-eQTLs cannot target all genes\n",
    "        choose = np.random.choice(ngene, ncausal, replace = False)\n",
    "        cindex[i, choose] = 1                                                    # mark these genes as causal\n",
    "\n",
    "    gtarget = list()\n",
    "\n",
    "    for i in range(ngene):\n",
    "        csnps = np.where(cindex[:, i] == 1)[0]\n",
    "        if csnps.shape[0] > 0: # then we got a trans-eQTL\n",
    "            betas = np.random.normal(0, beta, size = csnps.shape[0])\n",
    "            #betas *= np.sqrt( H2[i] / np.sum(np.square(beta)) )\n",
    "            liabilities[i, :] = np.dot(gt[csnps, :].T, betas)\n",
    "            gtarget.append(i)\n",
    "\n",
    "    newGX = gx + liabilities\n",
    "    newGX = (newGX - np.mean(newGX, axis = 1).reshape(-1, 1)) / np.std(newGX, axis = 1).reshape(-1, 1)\n",
    "    return newGX, nc\n",
    "### Simulation of data\n",
    "\n",
    "def simulate(sb, sx):\n",
    "    ngene = 1000\n",
    "    nsample = 338\n",
    "    nsnp = sx.shape[0]\n",
    "    \n",
    "    betas = np.random.normal(0, sb, ngene)\n",
    "    Y = np.zeros((ngene, nsample))\n",
    "    for i in range(ngene):\n",
    "        Y[i, :] = np.random.normal(0, 1, nsample)\n",
    "\n",
    "    Y = (Y - np.mean(Y, axis = 1).reshape(-1, 1)) / np.std(Y, axis = 1).reshape(-1, 1)\n",
    "    yTbeta = np.dot(Y.T, betas)\n",
    "    X = np.zeros((nsnp, nsample))\n",
    "    for i in range(nsnp):\n",
    "        error = np.random.normal(0, sx[i], nsample)\n",
    "        X[i, :] = yTbeta + error\n",
    "    return X, Y\n",
    "def read_expression(filename):\n",
    "    gene_names = list()\n",
    "    expression = list()\n",
    "    with open(filename, 'r') as genefile:\n",
    "        header = genefile.readline()\n",
    "        donorids = header.strip().split(\"\\t\")[1:]\n",
    "        for line in genefile:\n",
    "            linesplit = line.strip().split(\"\\t\")\n",
    "            expression.append(np.array(linesplit[1:], dtype=float))\n",
    "            gene_names.append(linesplit[0])\n",
    "    expression = np.array(expression)\n",
    "    return donorids, expression, gene_names\n",
    "\n",
    "\n",
    "'''def norm_binom(genotype, f):\n",
    "    genotype = (genotype - (2 * f)) / np.sqrt(2 * f * (1 - f))\n",
    "    return genotype\n",
    "'''\n",
    "\n",
    "def parse_geno (genotype_filename, sample_filename, startsnp, endsnp):                              #Read genotype here\n",
    "    ds        = DosageParser(genotype_filename, sample_filename, startsnp, endsnp)\n",
    "    dosage    = ds.dosage\n",
    "    snpinfo   = ds.snpinfo\n",
    "    donorids  = ds.sample_id\n",
    "    nsnps     = ds.nsnps\n",
    "    nsample   = ds.nsample\n",
    "    return dosage, snpinfo, donorids, nsnps, nsample\n",
    "\n",
    "                                                                                                    #quality check, matching\n",
    "def quality_check ( sampleids, donorids):\n",
    "    choose_ids     = [x for x in donorids if x in sampleids]\n",
    "    #dosage_indices = [i for i, x in enumerate(donorids)  if x in choose_ids]\n",
    "    #exprsn_indices = [i for i, x in enumerate(sampleids) if x in choose_ids]\n",
    "    dosage_indices = [donorids.index(i) for i in choose_ids]\n",
    "    exprsn_indices = [sampleids.index(i) for i in choose_ids]\n",
    "    return dosage_indices, exprsn_indices\n",
    "\n",
    "def read_distance(filename):\n",
    "    samplelist = list()\n",
    "    distances = list()\n",
    "    with open(filename, 'r') as mfile:\n",
    "        mfile.readline()\n",
    "        for line in mfile:\n",
    "            linestrip = line.strip().split(\"\\t\")\n",
    "            samplelist.append(linestrip[0])\n",
    "            distances.append(np.array(linestrip[1:],dtype=float))\n",
    "    return np.array(distances), samplelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### Simulation of data\n",
    "\n",
    "# ngenes = 1000\n",
    "# nsamples = 200\n",
    "# nsnps = 10\n",
    "# true_sigmabeta = 0.02#0.0006\n",
    "# nmin = 1\n",
    "# cfrac = 1\n",
    "# nmax = int(ngenes)\n",
    "# ncausal = np.random.randint(nmin, nmax, nsnps)\n",
    "# fmin = 0.1\n",
    "# fmax = 0.5\n",
    "# ntrans = 5\n",
    "\n",
    "# Y = np.random.randn(ngenes * nsamples).reshape(ngenes, nsamples)\n",
    "\n",
    "# gtnorm, gtcent = permuted_dosage(Y, nsnp = nsnps, fmin = fmin, fmax = fmax)\n",
    "       \n",
    "# # Trans-eQTL\n",
    "# if ntrans > 0:\n",
    "#     newgx,  nc = Expression(gtnorm[-ntrans:, :], Y, beta=true_sigmabeta, cfrac = cfrac)\n",
    "# Y = newgx\n",
    "\n",
    "# # X = np.zeros((nsnps, nsamples))\n",
    "# # for i in range(nsnps):\n",
    "# #     choose = np.random.choice(ngenes, ncausal[i], replace=False)\n",
    "# #     betas = np.random.normal(0, true_sigmabeta, ncausal[i])\n",
    "# #     #betas = np.random.normal(0, true_sigmabeta, ngenes)\n",
    "# #     #choose = np.arange(ngenes)\n",
    "# #     X[i, :] = np.dot(Y[choose, :].T, betas) + np.random.normal(0, 1, nsamples)\n",
    "\n",
    "# # sigmax = np.linspace(0, 0.4, nsnps)\n",
    "# # X, Y = simulate(true_sigmabeta, sigmax)\n",
    "# # gtcent = X - np.mean(X, axis = 1).reshape(-1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "==================================================\n",
      "low maf  poly alleles  complement alleles \n",
      "\n",
      "402   64   187 \n",
      "\n",
      "Total SNPs read in :  1000\n",
      "Total SNPs remained :  347\n",
      "==================================================\n",
      "\n",
      "Completed data loading and processing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#out_fileprefix      = \n",
    "genotype_filename   = \"data/geno/GTEx_450Indiv_genot_imput_info04_maf01_HWEp1E6_dbSNP135IDs_donorIDs_dosage_chr2.gz\"\n",
    "sample_filename     = \"data/geno/donor_ids.fam\"\n",
    "#expression_filename = \"/home/raktim/Documents/tejaas/final/PCA_KNN_Correction/main/expr2.txt\"\n",
    "expression_filename = \"Whole_Blood_Analysis.v6p.normalized.expression.txt\"\n",
    "\n",
    "startsnp            = 1\n",
    "endsnp              = 1000\n",
    "optim               = 0\n",
    "#user_sigbeta        = 0.001636581825407\n",
    "#transgeno           = \"data/GTEx_450Indiv_genot_imput_info04_maf01_HWEp1E6_dbSNP135IDs_donorIDs_dosage_filteredTEJAAS_transeqtls.gz\"#opts.trans_genofile\n",
    "sampleids, expression, gene_names = read_expression(expression_filename)\n",
    "dosage, snpinfo, donorids, nsnps, nsample = parse_geno ( genotype_filename, sample_filename, startsnp, endsnp)\n",
    "dosage_indices, exprsn_indices = quality_check (sampleids , donorids)\n",
    "conv_dosage = np.round(dosage)\n",
    "expr = expression[:, exprsn_indices]\n",
    "#expr = np.random.normal(0, 1, 23973 * 338).reshape(23973, 338)\n",
    "geno = conv_dosage[:, dosage_indices]\n",
    "Y = (expr - np.mean(expr, axis = 1).reshape(-1, 1)) / np.std(expr, axis = 1).reshape(-1, 1)\n",
    "Y = Y[:500]\n",
    "gtcent = geno - np.mean(geno, axis = 1).reshape(-1, 1)\n",
    "print (\"Completed data loading and processing\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# from scipy.optimize import minimize, fmin_l_bfgs_b\n",
    "# import regstat\n",
    "# from tqdm import tqdm\n",
    "# import qstats.crrstat as crrstat\n",
    "\n",
    "\n",
    "\n",
    "# class OptimizeRegularizer:\n",
    "\n",
    "#     def __init__(self, genotype, gene_expr, sigmax = 1, tol = 0.01, sigmabeta = 2):\n",
    "#         self._genotype = genotype\n",
    "#         self._expr = gene_expr\n",
    "#         self._sigmax = sigmax\n",
    "#         self._sigmareg = sigmabeta\n",
    "#         self._tolerance = tol\n",
    "#         self._niter = 0\n",
    "#         self._U, self._S, self._vt = self._svd(self._expr)\n",
    "\n",
    "#         self._A =  np.dot(self._expr,self._expr.T)\n",
    "#         self._detA = np.linalg.det(self._A)\n",
    "#         self._trA  = np.trace(self._A)\n",
    "#         self._trA_2 = np.trace(self._A)**2\n",
    "#         self._tr_A2 = np.trace(np.dot(self._A,self._A))\n",
    "\n",
    "#     def _svd(self, expr):\n",
    "#         U, S, Vt = np.linalg.svd(np.transpose(expr),full_matrices=False)\n",
    "#         return (U, S, Vt)\n",
    "\n",
    "#     @property\n",
    "#     def sigmareg(self):\n",
    "#         return self._sigmareg\n",
    "\n",
    "\n",
    "#     @property\n",
    "#     def niter(self):\n",
    "#         return self._niter\n",
    "\n",
    "#     def _logml(self, logsigmab, *args):\n",
    "#         sigmabeta = np.e ** logsigmab\n",
    "#         X, Y, S, U, sigmax = args\n",
    "#         nsnps      = X.shape[0]\n",
    "#         nsamples   = X.shape[1]\n",
    "#         ngenes     = Y.shape[0]\n",
    "#         sigmabeta2 = sigmabeta * sigmabeta\n",
    "#         sigmax2    = sigmax**2\n",
    "#         yty = np.dot(Y, Y.T)\n",
    "#         lml = 0\n",
    "#         for snp in range(nsnps):\n",
    "#             #A = yty.copy()\n",
    "#             #print(np.linalg.slogdet(A)[1])\n",
    "#             #A[np.diag_indices(ngenes)] += sigmax2[snp] / sigmabeta2\n",
    "#             #logdetA = np.linalg.slogdet(A)[1]\n",
    "#             #eps = sigmabeta2/sigmax2[snp]\n",
    "#             #log_identity_term = - np.log(eps) * ngenes\n",
    "#             #logdetA = log_identity_term + np.log(1 + self._detA + eps * self._trA + 0.5 * eps * eps * self._trA_2 - 0.5 * eps * eps *self._tr_A2)\n",
    "\n",
    "#             Smod = np.diag(np.square(S) / (np.square(S) + sigmax2[snp] / sigmabeta2))\n",
    "#             logdetA = np.sum(np.log(np.square(S) + sigmax2[snp] / sigmabeta2) ) + (ngenes - nsamples)*np.log(sigmax2[snp] / sigmabeta2)\n",
    "#             #print(logdetA)\n",
    "#             W = np.dot(U, np.dot(Smod, U.T))\n",
    "#             const_term = - 0.5 * ngenes * np.log(2 * np.pi * sigmabeta2) - 0.5 * logdetA\n",
    "#             snp_term = 0.5 * np.dot(X[snp, :].T, np.dot(W, X[snp,:])) / sigmax2[snp]\n",
    "#             lml = lml + (const_term) + snp_term\n",
    "#         #print(-lml)\n",
    "#         return -lml\n",
    "\n",
    "#     def _grad_logml(self, logsigmab, *args):\n",
    "#         sigmabeta = np.e ** logsigmab          #x is logsigmabeta\n",
    "#         X, Y, S, U, sigmax = args\n",
    "#         nsnps = X.shape[0]\n",
    "#         nsamples = X.shape[1]\n",
    "#         ngenes = Y.shape[0]\n",
    "#         sigmabeta2 = sigmabeta * sigmabeta\n",
    "#         sigmax2    = sigmax**2\n",
    "\n",
    "#         term1 = -ngenes\n",
    "\n",
    "\n",
    "\n",
    "#         der = 0\n",
    "#         for i in range(nsnps):\n",
    "#             Smod = np.square(S) * sigmabeta2 / sigmax2[i]\n",
    "#             term2 = (ngenes - nsamples) + np.sum(1 / (Smod + 1))\n",
    "#             term3 = 0\n",
    "#             for k in range(nsamples):\n",
    "#                 uk = U[:, k]\n",
    "#                 sk = S[k]\n",
    "#                 smod = sk * sk * sigmabeta2 / sigmax2[i]\n",
    "#                 term3 += smod * np.square(np.dot(uk, X[i,:])) / sigmax2[i] / np.square(smod + 1)\n",
    "#             der += term1 + term2 + term3\n",
    "#         return der\n",
    "\n",
    "#     def update (self):\n",
    "#         sigmareg_old = np.e ** self._sigmareg\n",
    "#         N = 10                                                                                         # expected number of true trans-eQTLs. Hard-coded\n",
    "#         iterate = True\n",
    "#         sigmax2 = self._sigmax ** 2\n",
    "#         logsb = 2\n",
    "#         arguments = ()\n",
    "#         def call(x):\n",
    "#             print(x)\n",
    "#         while iterate:\n",
    "#             Qsvd = np.array(range(self._genotype.shape[0]))\n",
    "#             #for snp in tqdm(range(self._genotype.shape[0])):\n",
    "#             #Q, _, _  = regstat.rscore(self._genotype[snp:snp+1,:], self._expr, sigmareg_old**2, np.array([sigmax2[snp]])) \n",
    "#             _, Q, _, _ = crrstat.perm_null(self._genotype, self._expr, np.array([sigmareg_old**2]*len(Qsvd)), sigmax2)\n",
    "#             Qsvd = Q\n",
    "#             #print(np.argsort(-Qsvd)[0:N])\n",
    "#             #print(Qsvd[0:N])\n",
    "#             top_Q_indx = np.argsort(-Qsvd)[0:N]\n",
    "\n",
    "#             top_geno = self._genotype[top_Q_indx,:]\n",
    "#             arguments = (top_geno, self._expr, self._S, self._U, self._sigmax[top_Q_indx])\n",
    "\n",
    "#             print(\"optimizing...\")\n",
    "#             #print(self._logml(self._sigmareg, *arguments))\n",
    "#             #res = fmin_l_bfgs_b(self._logml, [2], fprime = self._grad_logml, args = arguments)# jac=self._grad_logml)\n",
    "#             res = minimize(self._logml, [logsb], method=\"BFGS\", args = arguments, options={'disp': True}, callback=call)\n",
    "#             #res = minimize(self._logml, [20], method=\"nelder-mead\", args = arguments) \n",
    "#             #print(res)\n",
    "#             logsb = res.x[0]\n",
    "#             sigbeta_new = np.e**res.x[0]\n",
    "#             print(\"current sigmabeta\", sigbeta_new)\n",
    "\n",
    "#             checksigma = self.check_convergence(sigbeta_new, sigmareg_old)\n",
    "\n",
    "#             if checksigma:\n",
    "#                 iterate = False\n",
    "#                 sigma_optimized = sigbeta_new\n",
    "\n",
    "#             self._niter += 1\n",
    "#             sigmareg_old = sigbeta_new\n",
    "\n",
    "#         self._sigmareg = sigma_optimized\n",
    "#     def check_convergence(self, x, xold):\n",
    "#         check = False\n",
    "#         tol = self._tolerance\n",
    "#         diff = x - xold\n",
    "#         if diff == 0:\n",
    "#             check = True\n",
    "#         if not check and xold != 0.0:\n",
    "#             diff_percent = abs(diff) / xold\n",
    "#             if diff_percent < tol:\n",
    "#                 check = True\n",
    "\n",
    "#         return check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(500, 338)\n"
     ]
    }
   ],
   "source": [
    "print(Y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|          | 0/347 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Sigma beta optimization started. Reading from the provided trans-genotype file.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 347/347 [00:16<00:00, 20.62it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimized sigma beta value is: [2.09044492e-05 1.46892451e-03 2.75052179e-03 8.77675856e-03\n",
      " 5.72156549e-03 9.12740702e-03 3.17945328e-07 8.20740225e-03\n",
      " 5.60340018e-03 9.24990981e-03 8.39056209e-03 8.17524700e-03\n",
      " 1.63881913e-05 8.39056209e-03 9.71750347e-04 9.41995396e-03\n",
      " 6.90568012e-03 2.94999633e-05 8.98434730e-03 4.97918069e-03\n",
      " 8.01672505e-03 8.03212302e-03 1.04068145e-02 1.04068145e-02\n",
      " 2.54751691e-03 3.22537431e-05 1.04068145e-02 1.04382682e-02\n",
      " 3.30621070e-05 1.82796328e-05 1.63025474e-05 6.36065578e-03\n",
      " 1.89091663e-03 1.79420302e-03 8.31279516e-04 5.26333282e-04\n",
      " 6.50290053e-03 2.17519644e-05 2.31305779e-03 2.56359183e-05\n",
      " 5.12886885e-03 2.56359183e-05 2.71599480e-05 2.56359183e-05\n",
      " 5.87837500e-03 2.71599480e-05 5.78579350e-03 5.12886885e-03\n",
      " 1.45709966e-06 7.30387979e-03 6.10462830e-03 3.10961214e-05\n",
      " 2.28383078e-05 5.44623218e-03 6.10462830e-03 1.75642483e-05\n",
      " 6.10462830e-03 6.10462830e-03 4.27370978e-03 6.10462830e-03\n",
      " 5.51762093e-03 6.10462830e-03 2.51290741e-05 6.10462830e-03\n",
      " 1.88826650e-03 2.87739505e-05 1.88826650e-03 1.88826650e-03\n",
      " 6.12140134e-03 1.88826650e-03 5.51762093e-03 6.10462830e-03\n",
      " 7.65110120e-03 5.51762093e-03 1.88826650e-03 5.61008505e-03\n",
      " 5.51762093e-03 6.10118450e-05 6.10462830e-03 6.04292653e-03\n",
      " 5.77338447e-03 6.70701553e-03 2.61819291e-03 1.80666012e-22\n",
      " 6.70701553e-03 5.66770416e-03 4.85454377e-05 4.85454377e-05\n",
      " 1.16767809e-02 4.75077648e-03 7.19178889e-03 9.03681197e-03\n",
      " 1.63236715e-05 1.47500854e-05 5.30997706e-03 5.01700576e-03\n",
      " 4.06229657e-03 5.01700576e-03 5.01700576e-03 4.06229657e-03\n",
      " 5.01700576e-03 1.49802276e-07 1.72366560e-05 1.72366560e-05\n",
      " 5.01700576e-03 1.49802276e-07 1.05372721e-08 4.46418375e-03\n",
      " 5.84582486e-03 6.58199921e-03 4.06229657e-03 5.01700576e-03\n",
      " 4.06229657e-03 3.72772491e-05 4.54127310e-03 1.09298236e-04\n",
      " 3.60950659e-03 4.06229657e-03 5.01700576e-03 1.72366560e-05\n",
      " 6.87794302e-03 5.01700576e-03 1.72366560e-05 3.59803368e-05\n",
      " 5.52379616e-03 7.44558590e-03 4.06229657e-03 5.01700576e-03\n",
      " 5.01700576e-03 1.49802276e-07 5.01700576e-03 4.36710079e-03\n",
      " 5.01700576e-03 2.53009538e-06 1.37332023e-02 5.08945350e-03\n",
      " 5.08945350e-03 4.85412607e-03 1.15454000e-06 8.19665503e-06\n",
      " 1.49802276e-07 5.08945350e-03 4.09766109e-03 5.08945350e-03\n",
      " 5.08945350e-03 7.65366210e-03 4.36710079e-03 4.36710079e-03\n",
      " 5.08945350e-03 5.08945350e-03 8.19665503e-06 4.36710079e-03\n",
      " 5.93577895e-03 6.19041389e-03 5.93577895e-03 4.36710079e-03\n",
      " 5.30819401e-03 8.50020910e-06 8.15805279e-03 4.97970931e-03\n",
      " 5.08945350e-03 6.25951833e-05 5.08945350e-03 4.36710079e-03\n",
      " 7.65366210e-03 3.54960338e-07 1.99278967e-19 1.72366560e-05\n",
      " 5.14047894e-03 5.08945350e-03 7.65366210e-03 1.72366560e-05\n",
      " 8.19665503e-06 4.69111885e-03 4.36710079e-03 7.65366210e-03\n",
      " 3.10344941e-05 3.99092228e-03 1.38983549e-03 7.26858755e-03\n",
      " 6.13658713e-03 4.38346485e-03 4.38346485e-03 4.38346485e-03\n",
      " 9.63140828e-05 1.47035300e-05 2.48402839e-05 6.21020130e-03\n",
      " 4.38346485e-03 1.47035300e-05 1.47035300e-05 9.63140828e-05\n",
      " 9.63140828e-05 3.93393371e-03 5.54608556e-03 4.92890797e-06\n",
      " 3.94697657e-03 5.63141064e-03 6.49084523e-03 5.52754531e-03\n",
      " 1.20403094e-02 1.20403094e-02 3.12391746e-03 6.49084523e-03\n",
      " 3.49278131e-03 1.19549846e-02 4.59679929e-06 3.49278131e-03\n",
      " 6.56446026e-05 7.62612271e-05 5.35698085e-06 5.35698085e-06\n",
      " 3.50126931e-05 1.85381691e-07 1.15454000e-06 1.30318803e-02\n",
      " 6.49444997e-06 1.21065063e-03 4.22103500e-03 6.05894835e-03\n",
      " 2.11385280e-05 6.85008300e-03 8.51348664e-07 3.37748739e-05\n",
      " 6.93803174e-03 7.04084834e-03 7.15402505e-03 2.48452870e-03\n",
      " 1.10121558e-02 6.36411471e-03 9.40448166e-03 1.63904105e-02\n",
      " 2.48452870e-03 1.10121558e-02 1.10121558e-02 1.10121558e-02\n",
      " 5.88823159e-03 1.65653486e-02 5.88823159e-03 3.50858282e-07\n",
      " 9.78903423e-03 1.41182042e-03 1.41182042e-03 6.84328026e-03\n",
      " 1.10121558e-02 8.20846808e-03 8.20846808e-03 1.09039391e-03\n",
      " 1.67211654e-02 2.99644749e-03 1.07363169e-02 1.10121558e-02\n",
      " 9.90358937e-03 8.42024950e-06 1.76485018e-02 1.72867602e-02\n",
      " 1.65102639e-05 5.99010498e-03 1.18889937e-02 1.18889937e-02\n",
      " 1.18889937e-02 1.91632052e-03 1.56241815e-02 7.63691051e-03\n",
      " 1.01007397e-03 3.24686165e-03 5.12062189e-03 3.06250429e-06\n",
      " 1.67709748e-09 1.33408924e-05 1.00424603e-05 1.72086896e-02\n",
      " 7.03355168e-06 1.09509353e-02 1.55222072e-02 1.09534997e-02\n",
      " 1.09534997e-02 4.23246822e-03 7.03355168e-06 1.12210700e-02\n",
      " 1.12210700e-02 1.00416898e-05 5.94934230e-03 1.05128398e-02\n",
      " 2.35646285e-03 1.59960894e-02 1.08775176e-03 1.29793900e-05\n",
      " 1.47406300e-03 1.67079732e-05 7.76760295e-27 1.20553204e-03\n",
      " 1.55603382e-03 1.19481052e-05 1.70630937e-05 8.99504402e-03\n",
      " 1.32288150e-06 2.54095477e-03 7.12474016e-07 6.29609958e-04\n",
      " 6.51016487e-08 7.11714111e-06 9.11387080e-06 9.39868070e-06\n",
      " 4.48928561e-08 4.34117500e-06 2.39889063e-05 4.34117500e-06\n",
      " 1.87271977e-05 2.46918958e-06 4.88325836e-05 1.03580208e-79\n",
      " 3.46598426e-06 2.36483893e-05 2.36483893e-05 8.33079362e-06\n",
      " 3.46598426e-06 1.51508298e-05 1.10904736e-06 1.41769395e-02\n",
      " 3.25503235e-05 8.58644427e-32 3.46598426e-06 3.83886109e-05\n",
      " 3.46598426e-06 3.46598426e-06 3.46598426e-06 3.46598426e-06\n",
      " 6.02320557e-07 9.75326997e-06 7.88078856e-03 3.46598426e-06\n",
      " 6.93951770e-03 1.93942651e-05 2.42949190e-05 1.04229488e-02\n",
      " 1.99120439e-05 3.46598426e-06 9.75326997e-06 3.97404039e-03\n",
      " 3.03155665e-05 1.11910870e-02 1.84030957e-07 1.62098387e-05\n",
      " 6.25597182e-03 1.62098387e-05 1.62098387e-05] \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "optim = True\n",
    "sigbeta = 0.6\n",
    "if optim:\n",
    "    print (\"\\nSigma beta optimization started. Reading from the provided trans-genotype file.\")\n",
    "    try:\n",
    "        sigmax = np.std(gtcent - np.mean(gtcent, axis = 0),axis = 1)\n",
    "        #optimize_sigbeta   = OptimizeRegularizer(geno[:10], expr, sigmax = np.std(geno[:10].T,axis = 0))\n",
    "        optimize_sigbeta   = OptimizeRegularizer(gtcent, Y, sigmax = sigmax)\n",
    "        l = optimize_sigbeta.update()\n",
    "        optimize_sigbeta.niter\n",
    "        sigbeta            = optimize_sigbeta.sigmareg\n",
    "        #print(sigbetas)\n",
    "        \n",
    "        print (\"Optimized sigma beta value is:\" , np.array(sigbeta),\"\\n\")\n",
    "        #del geno, conv_dosage, trans_dosage\n",
    "    except OSError as e:\n",
    "        print(\"\\n\",e, \". Trans-eQTLs file not provided for Optimization\\n\")\n",
    "        raise SystemExit\n",
    "\n",
    "\n",
    "else:\n",
    "    print(\"\\n=============================================================\")\n",
    "    print(\"\\nsigma beta optimization not requested\")\n",
    "    if user_sigbeta == 0.006:\n",
    "        print (\"\\nsigma beta not provided. Using default value of 0.006\")\n",
    "    else:\n",
    "        print(\"\\nsigma beta value provided as : \" , user_sigbeta)\n",
    "    print(\"=============================================================\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Marginal log likelihood (different ways of calculation)\n",
    "sigmax = np.std(gtcent,axis = 1)\n",
    "optimize_sigbeta   = OptimizeRegularizer(gtcent, Y, sigmax)\n",
    "\n",
    "sigmabeta = np.logspace(-10, 4, 50, base = np.e)\n",
    "\n",
    "U, S, Vt = np.linalg.svd(Y.T)\n",
    "\n",
    "A =  np.dot(Y, Y.T)\n",
    "detA = np.linalg.det(A)\n",
    "trA  = np.trace(A)\n",
    "trA_2 = np.trace(A)**2\n",
    "tr_A2 = np.trace(np.dot(A,A))\n",
    "args = (gtcent, Y, S, U, sigmax)\n",
    "ysimp    = [optimize_sigbeta._logml(np.log(sbeta), *args) for sbeta in sigmabeta]\n",
    "ycomp    = [logml(gtcent, Y, sbeta, sigmax) for sbeta in sigmabeta]\n",
    "#ycompsvd = [logml_appr_old(gtcent, Y, sbeta, sigmax, detA, trA, trA_2, tr_A2) for sbeta in sigmabeta]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigmabeta = np.logspace(-10, 4, 50, base = np.e)\n",
    "true_alpha = np.log(true_sigmabeta)\n",
    "#print(true_sigmabeta)\n",
    "\n",
    "fig = plt.figure(figsize = (6, 6))\n",
    "ax1 = fig.add_subplot(111)\n",
    "\n",
    "ax1.scatter(np.log(sigmabeta), -np.array(ysimp),    color = 'blue', label = 'Optim_lml')\n",
    "ax1.plot(np.log(sigmabeta),    ycomp,    color = 'green', label = 'Complex')\n",
    "#ax1.scatter(np.log(sigmabeta), ycompsvd, color = 'salmon', label = 'Complex SVD')#'salmon'\n",
    "print(true_alpha)\n",
    "#ax1.axvline(x = true_alpha, color = 'red', label = 'True')\n",
    "\n",
    "\n",
    "ax1.axvline(x = np.log(sigbeta), color = 'blue', label = 'Optim_result')\n",
    "\n",
    "plt.legend(loc = 'upper left')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "U, S, Vt = np.linalg.svd(np.transpose(Y),full_matrices=False)\n",
    "print(Y.shape)\n",
    "print(S.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "YtY = np.dot(Y, Y.T)\n",
    "print(YtY.shape)\n",
    "YtY[np.diag_indices(Y.shape[0])] += 5 #sigmax2[snp] / sigmabeta2\n",
    "\n",
    "print(np.linalg.slogdet(YtY))\n",
    "#print(np.product(S**2))\n",
    "print(np.sum(np.log(S**2 + 5)) + (Y.shape[0] - Y.shape[1])*np.log(5) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
